patch/gfx9/llpcNggPrimShader.cpp

```diff
@@ -1432,7 +1433,10 @@ void NggPrimShader::ConstructPrimShaderWithoutGs(
 
                 // The DWORD following DWORDs for all waves stores the vertex count of the entire sub-group
                 Value* pLdsOffset = m_pBuilder->getInt32(regionStart + waveCountInSubgroup * SizeOfDword);
-                pVertCountInWaves = m_pLdsManager->ReadValueFromLds(m_pBuilder->getInt32Ty(), pLdsOffset);
+                // NOTE: Use volatile load to prevent LLVM optimization because CFG simplification will remove this
+                // block and use select instruction to select the result between LDS load value and zero. This is not
+                // our expectation.
+                pVertCountInWaves = m_pLdsManager->ReadValueFromLds(m_pBuilder->getInt32Ty(), pLdsOffset, true);
```

Comment:
Do you have a sample of the relevant LLVM IR and resulting ISA and why you think this is actually worse code? It would be good to get the backend to do the right thing here rather than hack around it, and we should definitely understand it better...

Also, your commit message mentions assertions in LLVM. What are those, and do you have LLVM IR to reproduce them?

Reply:
Yes, I will two samples to reproduce the issues: one is a bad code generation (I find volatile LDS load could prevent this optimization), the other is a assert in scheduler (saying scheduling failed, I find this is because I set "volatile" to LDS load). I can file issue tickets to backend team to further check them.

Reply:
Thank you, please add me as a watcher when you do so.

Reply:
I just file an issue (LLVM assert). For another issue of bad code generation, I double check it. The semantics are correct, not a problem as I initially believed. I will correct my commit. Thanks for your helpful review :)